packages/jest-cli/src/TestRunner.js

```diff
@@ -312,12 +311,11 @@ class TestRunner {
       if (watcher.isWatchMode()) {
         watcher.setState({interrupted: true});
       } else {
-        this._dispatcher.onRunComplete(
-          contexts,
-          this._globalConfig,
-          aggregatedResults,
-        );
-        process.exit(1);
+        const exit = () => process.exit(1);
+        this._dispatcher
```

Comment:
This being async will not work as it will keep running tests until the reporters are done. You need to wait on the result of _bailIfNeeded and return `Promise.resolve()` otherwise. I'll send a fix.